git-user-count/user-count.sh

```diff
@@ -0,0 +1,144 @@
+#!/bin/bash
+set -e
+
+basedir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+# input
+repositories="$basedir/repositories.txt"
+
+# outputs
+userListFile="$basedir/gradle-enterprise-users.txt"
+userUniqueListFile="$basedir/gradle-enterprise-unique-users.txt"
+usersByRepoFile="$basedir/gradle-enterprise-users-by-repo.csv"
+checkout_area=.tmp/repos
+
+# options
+since=30
+shallow_clone=true
+git_options=
+
+yellow='\033[1;33m'
+nc='\033[0m'
+
+# -b option lets you specify a branch to checkout, store it in a variable
+# -s option lets you specify a number of days to count commits from, store it in a variable otherwise default to 30
+# -c option disables shallow cloning
+# -o option lets you specify aditional git cloning options
+while getopts b:s:co: option
+do
+  case "${option}" in
+    b) branch=${OPTARG};;
+    s) since=${OPTARG};;
+    c) shallow_clone=false;;
+    o) git_options=${OPTARG};;
+    *) echo "Usage: $0 [-b branch_name] [-s since_days] [-o git_options] [-c]" >&2
+       exit 1;;
+  esac
+done
+
+# Check that both -s and -c cannot be specified at the same time
+if [ "$shallow_clone" = false ] && [ -n "$since" ]; then
+  echo "Cannot specify both -s and -c options at the same time" >&2
+  exit 1
+fi
+
+function prepare() {
+  echo -n "" > "$userListFile"
+  echo -n "" > "$userUniqueListFile"
+  echo "repository,users" > "$usersByRepoFile"
+}
+
+function print_information() {
+  echo "Counting commits from the last $since days"
+
+  if [ -z "$branch" ]; then
+    echo "No branch name specified, counting commits on default branch"
+  else
+    echo "Branch name is $branch"
+  fi
+}
+
+function process_repositories() {
+  numOfRepos=$(sed "/^[[:blank:]]*#/d;s/^[[:blank:]]*//;s/#.*//" "$repositories" | wc -l | tr -d '[:space:]')
+  current=1
+  while IFS= read -r repo
+  do
+    echo -e -n "${yellow}($current/$numOfRepos) Processing ${repo}...${nc}"
+    process_repository "$repo"
+    ((current++))
+  done < <(sed "/^[[:blank:]]*#/d;s/^[[:blank:]]*//;s/#.*//" "$repositories")
+
+  # remove duplicates in list of users
+  sort -u "$userListFile" > "$userUniqueListFile"
+}
+
+function process_repository() {
+  repository_name="${1//\//-}"
+  
+  # Check if repository is valid
+  if [ -z "$repository_name" ]; then
+    echo "repository name is empty" >&2
+  # If repository is already cloned, switch to specified branch
+  elif [ -d "$checkout_area/$repository_name" ]; then
+    echo "already cloned."
```

Comment:
It might be a good idea to update the existing clone:

```
git pull
```

Otherwise, the script will only find data from when the clone was first created.

Reply:
I think an option to decide on the behavior could make sense, the current implementation allows to resume from a previous run. This can be very helpful when encountering an error with a large list of repos.

Reply:
These are both good points. This logic was originally intented to be used as Jerome wrote - to resume from a previous run. I would think this is the majority of the use cases here. I don't imagine that someone would keep all these repos around and then run them months later.